baqend/_getFMPData.ts

```diff
@@ -0,0 +1,209 @@
+
+  if (!matchArray || !matchArray[1]) {
+    throw new Error('Could not find data table in HTML.')
+  }
+
+  const data = JSON.parse(matchArray[1].replace(/'/g, '"')) as Array<[string, number]>
+
+  // Remove the first item of the data array because it has irrelevant data like "Time (ms)"
+  data.shift()
+
+  return data.map(row => [parseFloat(row[0]), row[1]] as [number, number])
+}
+
+/**
+ * Gets the FMP calculated by WebPagetest.
+ */
+function getFMPFromWebPagetest(data: WptView): number {
+  // Search First Meaningful Paint from timing
+  const { chromeUserTiming = [] } = data
+  const firstMeaningfulPaintObject =
+    chromeUserTiming
+      .reverse()
+      .find(entry => entry.name === 'firstMeaningfulPaint' || entry.name === 'firstMeaningfulPaintCandidate')
+
+  return firstMeaningfulPaintObject ? firstMeaningfulPaintObject.time : 0
+}
+
+/**
+ * Calculates the first deviation of the given data.
+ */
+function firstDeviation(db: baqend, data: Array<[number, number]>): model.Candidate[] {
+  const diffs = [] as model.Candidate[]
+  if (data.length === 1) {
+    const candidate = new db.Candidate()
+    candidate.visualCompleteness = 100
+    candidate.deltaVC = data[0][1]
+    candidate.startTime = 0
+    candidate.endTime = data[0][0] * 1000
+    candidate.wptFMP = null
+    return candidate
+  }
+
+  let lastVisualProgress = data[0][1]
+  for (let i = 1; i < data.length; i += 1) {
+    const [time, visualProgress] = data[i]
+    const diff = visualProgress - lastVisualProgress
+    lastVisualProgress = visualProgress
+
+    const candidate = new db.Candidate()
+    candidate.visualCompleteness = visualProgress
+    candidate.deltaVC = diff
+    candidate.startTime = time * 1000 - 100
+    candidate.endTime = time * 1000
+    candidate.wptFMP = null
+    diffs.push(candidate)
+  }
+
+  return diffs
+}
+
+/**
+ * Calculates ΔVisualCompleteness from the WebPagetest
+ */
+async function prepareCandidates(db: baqend, testId: string, runIndex: string): Promise<model.Candidate[]> {
+  const url = `http://${credentials.wpt_dns}/video/compare.php?tests=${testId}-r:${runIndex}-c:0`
+  const response = await fetch(url)
+  const htmlString = await response.text()
+  const data = getDataFromHtml(htmlString)
+  db.log.info('Found data for FMP calculation', { data })
+
+  return firstDeviation(db, data)
+}
+
+/**
+ * Chooses up to five candidates based on the given deltas.
+ */
+function chooseTopCandidates(db: baqend, candidates: model.Candidate[]): model.Candidate[] {
+  // Find the five highest ΔVCs
+  const topCandidates = candidates
+    .sort(({ deltaVC: a }, { deltaVC: b }) => b - a)
+    .slice(0, 5)
+
+  db.log.info('Candidates for FMP found', { topCandidates })
+  return topCandidates
+}
+
+/**
+ * Gets a candidate that matches the WPT FMP or null.
+ */
+function getValidCandidate(wptFMP: number, candidates: model.Candidate[]): model.Candidate|null {
+  if (wptFMP <= 0) {
+    return null
+  }
+
+  if (candidates.length === 1) {
```

Comment:
Can we remove the whole length === 1 case as it is covered by find method anyway?